raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java

```diff
@@ -101,43 +107,66 @@ public void punctuate(long timestamp) {
       return;
     }
 
+    long emitTs = traceState.getEmitTs();
     if (emitTs <= timestamp) {
       // we can emit this trace so just delete the entry for this 'key'
       traceEmitTriggerStore.delete(key);
       // Implies that no new spans for the trace have arrived within the last
       // 'groupingWindowTimeoutMs' interval
       // so the trace can be finalized and emitted
-      ValueAndTimestamp<RawSpans> agg = inflightTraceStore.delete(key);
-      ByteBuffer traceId;
-      String tenantId;
-      if (agg != null && !agg.value().getRawSpans().isEmpty()) {
-        RawSpan firstSpan = agg.value().getRawSpans().get(0);
-        traceId = firstSpan.getTraceId();
-        tenantId = firstSpan.getCustomerId();
-
-        List<RawSpan> rawSpanList = agg.value().getRawSpans();
-        recordSpansPerTrace(rawSpanList.size(), List.of(Tag.of("tenant_id", tenantId)));
-        Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
-        StructuredTrace trace =
-            StructuredTraceBuilder.buildStructuredTraceFromRawSpans(
-                rawSpanList, traceId, tenantId, timestamps);
 
-        logger.debug(
-            "Emit tenant_id=[{}], trace_id=[{}], spans_count=[{}]",
-            tenantId,
-            HexUtils.getHex(traceId),
-            rawSpanList.size());
-
-        tenantToTraceEmittedCounter
-            .computeIfAbsent(
-                tenantId,
-                k ->
-                    PlatformMetricsRegistry.registerCounter(
-                        TRACES_EMITTER_COUNTER, Map.of("tenantId", k)))
-            .increment();
-
-        context.forward(null, trace, outputTopicProducer);
-      }
+      ByteBuffer traceId = traceState.getTraceId();
+      String tenantId = traceState.getTenantId();
+
+      TreeSet<ByteBuffer> set =
+          new TreeSet<>(
+              (v1, v2) -> {
+                Serde<SpanIdentity> serde = (Serde<SpanIdentity>) context.keySerde();
+                byte[] b1 =
+                    serde.serializer().serialize("", new SpanIdentity(tenantId, traceId, v1));
+                byte[] b2 =
+                    serde.serializer().serialize("", new SpanIdentity(tenantId, traceId, v2));
+                return ByteBuffer.wrap(b1).compareTo(ByteBuffer.wrap(b2));
+              });
+      set.addAll(traceState.getSpanIds());
+
+      KeyValueIterator<Windowed<SpanIdentity>, RawSpan> iterator =
+          spanWindowStore.fetch(
+              new SpanIdentity(tenantId, traceId, set.first()),
+              new SpanIdentity(tenantId, traceId, set.last()),
+              Instant.ofEpochMilli(traceState.getTraceStartTimestamp()),
+              Instant.ofEpochMilli(traceState.getTraceEndTimestamp()));
+      List<RawSpan> rawSpanList = new ArrayList<>();
+      iterator.forEachRemaining(
+          v -> {
+            // range search could returns extra spans as well, filter them
```

Comment:
can u give an example ?

Reply:
Added example case in the doc, so I am passing low & high range of spanIdentifier in the query and keys are stored in a sorted manner in rocksdb. 
So it could happen when spanIdentifier for some other trace falls in the range of spanIdentifiers requested in the query.